geoips/utils/types/datatree_ditto.py

```diff
@@ -0,0 +1,600 @@
+
+    @_convert_output_to_DataTreeDitto
+    def __getitem__(self, key: str) -> Union["DataTreeDitto", Any]:
+        """Override getitem to return DataTreeDitto instances.
+
+        Parameters
+        ----------
+        key : str
+            Name of the child node to retrieve.
+
+        Returns
+        -------
+        DataTreeDitto
+            The child node, converted to DataTreeDitto if necessary.
+
+        Examples
+        --------
+        >>> dt = DataTreeDitto()
+        >>> dt["child"] = np.array([1, 2, 3])
+        >>> child = dt["child"]
+        >>> isinstance(child, DataTreeDitto)
+        True
+        """
+        return super().__getitem__(key)
+
+    # @classmethod
+    # def from_dict(cls, data=None, name=None):
+    #    # Pass keyword arguments directly, without the '*'
+    #    result = super().from_dict(data, name=name)
+    #    if isinstance(result, DataTree) and not isinstance(result, DataTreeDitto):
+    #        return cls._convert_datatree_to_ditto(result)
+    #    return result
+
+    @_convert_output_to_DataTreeDitto
+    def map_over_datasets(
+        self,
+        func: Callable[..., Any],
+        *args: Any,
+        kwargs: Mapping[str, Any] | None = None,
+    ) -> DataTree | tuple[DataTree, ...]:
+        return super().map_over_datasets(func, *args, kwargs=kwargs)
+
+    @_convert_output_to_DataTreeDitto
+    def _unary_op(self, f, *args, **kwargs) -> DataTree:
+        return super()._unary_op(f, *args, **kwargs)
+
+    @_convert_output_to_DataTreeDitto
+    def _binary_op(self, other, f, reflexive=False, join=None) -> DataTree:
+        return super()._binary_op(other, f, reflexive, join)
+
+    @_convert_output_to_DataTreeDitto
+    def filter(self: DataTree, filterfunc: Callable[[DataTree], bool]) -> DataTree:
+        return super().filter(filterfunc)
+
+    @_convert_output_to_DataTreeDitto
+    def match(self, pattern: str) -> DataTree:
+        return super().match(pattern)
+
+    @_convert_output_to_DataTreeDitto
+    def mean(
+        self,
+        dim=None,
+        *,
+        skipna=None,
+        keep_attrs=None,
+        **kwargs,
+    ):
+        return super().mean(dim, skipna=skipna, keep_attrs=keep_attrs, **kwargs)
+
+    @staticmethod
+    def _convert_datatree_to_ditto(dt: DataTree) -> "DataTreeDitto":
+        """Convert a DataTree to DataTreeDitto recursively.
+
+        Parameters
+        ----------
+        tt : DataTree
+            DataTree instance to convert.
+
+        Returns
+        -------
+        DataTreeDitto
+            Converted DataTreeDitto with all children also converted.
+        """
+        new_ditto = DataTreeDitto(dataset=dt.ds, name=dt.name)
+        for child_name, child in dt.children.items():
+            new_ditto[child_name] = DataTreeDitto._convert_datatree_to_ditto(child)
+        return new_ditto
+
+    def get_original(self, path: str = ".") -> Any:
+        """Get the original object (before conversion) at the specified path.
```

Comment:
This would get a new object that looks similar to the original but with potentially modified data, correct?

Reply:
This returns a functionally original object HOWEVER that "same" object has all of the same caveats of the original object. Namely if you modify a shallow copy the "original" data will also be modified.